app/grandchallenge/algorithms/signals.py

```diff
@@ -2,56 +2,79 @@
 from django.dispatch import receiver
 from guardian.shortcuts import assign_perm, remove_perm
 
-from grandchallenge.algorithms.models import Result
-from grandchallenge.cases.models import Image
+from grandchallenge.algorithms.models import Job
+from grandchallenge.components.models import ComponentInterfaceValue
 
 
-@receiver(m2m_changed, sender=Result.images.through)
-def update_image_permissions(instance, action, reverse, model, pk_set, **_):
+@receiver(m2m_changed, sender=Job.inputs.through)
+@receiver(m2m_changed, sender=Job.outputs.through)
+def update_input_image_permissions(
+    sender, instance, action, reverse, model, pk_set, **_
+):
     """
-    Assign or remove view permissions to the algorithms editors when images
-    are added or remove to/from the algorithm results. Handles reverse
+    Assign or remove view permissions to the algorithms editors when inputs
+    are added or remove to/from the algorithm jobs. Handles reverse
     relations and clearing.
     """
     if action not in ["post_add", "post_remove", "pre_clear"]:
         # nothing to do for the other actions
         return
 
+    if sender._meta.label_lower == "algorithms.job_inputs":
+        forward_lookup = "inputs"
+        reverse_lookup = "algorithms_job_inputs"
+    elif sender._meta.label_lower == "algorithms.job_outputs":
+        forward_lookup = "outputs"
+        reverse_lookup = "algorithms_job_outputs"
+    else:
+        raise RuntimeError("m2m is only valid for Job inputs and outputs.")
+
     if reverse:
-        images = Image.objects.filter(pk=instance.pk)
+        component_interface_values = ComponentInterfaceValue.objects.filter(
+            pk=instance.pk
+        )
         if pk_set is None:
             # When using a _clear action, pk_set is None
             # https://docs.djangoproject.com/en/2.2/ref/signals/#m2m-changed
-            algorithm_results = instance.algorithm_results.all()
+            jobs = getattr(instance, reverse_lookup).all()
         else:
-            algorithm_results = model.objects.filter(pk__in=pk_set)
+            jobs = model.objects.filter(pk__in=pk_set)
 
-        algorithm_results = algorithm_results.select_related(
-            "job__creator", "job__algorithm_image__algorithm__editors_group"
+        jobs = jobs.select_related(
+            "creator", "algorithm_image__algorithm__editors_group"
         )
     else:
-        algorithm_results = [instance]
+        jobs = [instance]
         if pk_set is None:
             # When using a _clear action, pk_set is None
             # https://docs.djangoproject.com/en/2.2/ref/signals/#m2m-changed
-            images = instance.images.all()
+            component_interface_values = getattr(
+                instance, forward_lookup
+            ).all()
```

Comment:
couldn't you filter these for `image__isnull=False`? That way you'd have a smaller queryset and wouldn't need the if statement on line 69.

Reply:
Great, thanks!